shoop/admin/modules/orders/views/shipment.py

```diff
@@ -17,15 +19,71 @@
 from shoop.admin.toolbar import PostActionButton, Toolbar
 from shoop.admin.utils.forms import add_form_errors_as_messages
 from shoop.admin.utils.urls import get_model_url
+from shoop.apps.provides import get_provide_objects
 from shoop.core.excs import NoProductsToShipException
 from shoop.core.models import Order, Product, Supplier
+from shoop.core.utils.shipments import generate_shipment_key
+
+FORM_MODIFIER_PROVIDER_KEY = "admin_extend_create_shipment_form"
+
+
+class ShipmentFormModifier(six.with_metaclass(abc.ABCMeta)):
+    @abc.abstractmethod
+    def get_extra_fields(self, order):
+        """
+        Extra fields for shipment creation form
+
+        :param order: Order linked to form
+        :type order: shoop.core.models.Order
+        :return: List of extra fields that should be added to form.
+        Tuple should contain field name and Django form field.
+        :rtype: list(tuple)|None
+        """
+        pass
+
+    @abc.abstractmethod
+    def clean_hook(self, cleaned_data):
+        """
+        Extra clean for shipment creation form
+
+        :param cleaned_data: form cleaned data
+        :type cleaned_data: dict
+        :return: List of errors that should be handled.
+        Tuple should contain field name and error message.
+        :rtype: list(tuple)|None
+        """
+        pass
+
+    @abc.abstractmethod
+    def form_valid_hook(self, form, order, shipment_key):
+        """
+        Extra form valid handler for shipment creation form
+
+        :param form: Form that is currently handled
+        :type form: ShipmentForm
+        :param order: Order linked to form
+        :type order: shoop.core.models.Order
+        :param shipment_key: Generated shipment key for shipment
+        that is created when all form valid operations is done.
+        :rtype: None
+        """
+        pass
```

Comment:
Document when this is called in the shipment creation process. (After creating shipment? Before invoking the signal?)